torch/distributed/distributed_c10d.py

```diff
@@ -562,7 +562,7 @@ def _store_based_barrier(rank, store, group_name, rendezvous_count, timeout, log
             # the status of the group or time out officially, throwing runtime error
             store.wait([last_worker_key], logging_interval)
             break
-        except RuntimeError as e:
+        except Exception as e:
```

Comment:
IIRC this code was written to support `store` instances of [`c10d::Store`](https://github.com/pytorch/pytorch/blob/main/torch/_C/_distributed_c10d.pyi#L150) (e.g. `torch.distributed.TCPStore`). 

On the job that raises the `StoreTimeoutHandlerError`, is the process group being initialized as:

```
dist.init_process_group(backend=backend, init_method="zeus://...")
```

by any chance? 

If so, we moved away from directly using zeus to initialize the process group a while ago. 

The solution is to invoke the main script with [`torchrun`](https://pytorch.org/docs/stable/elastic/run.html) with the `--rdzv_backend = "zeus"` configuration. 
You'd also have to change the `init_process_group` call-site to remove the `init_method` argument (`torchrun` will set the correct env vars in such a way that the process group can be initialized correctly with the default `init_method`).

In this case, zeus is used at the "node" level to perform peer discovery, but the actual trainers (those who invoke `init_process_group()`) will internally use `TCPStore`, therefore when `pg.barrier()` is called, the barrier would use `type(store) == TCPStore` and the code will do the right thing.

One improvement we could make is to actually define and pybind a `TimeoutError` from the store interface so that we can catch a narrower exception but that's a much larger change.

Currently (and I admit this isn't the best way) the pattern folks have used to catch timeout errors from the c10d store is by checking the RuntimeError message as done [here]( https://github.com/pytorch/pytorch/blob/main/torch/distributed/elastic/utils/distributed.py#L100)
which is better than checking the wall-clock time interval to determine a timeout as in L574 in this file.